src/domain/datasets/services/src/dataset_entry_service.rs

```diff
@@ -0,0 +1,285 @@
+        }: &DatasetLifecycleMessageCreated,
+    ) -> Result<(), InternalError> {
+        let entry = DatasetEntry::new(
+            dataset_id.clone(),
+            owner_account_id.clone(),
+            message.dataset_name().into_owned(),
+            self.time_source.now(),
+        );
+
+        self.dataset_entry_repo
+            .save_dataset_entry(&entry)
+            .await
+            .int_err()
+    }
+
+    async fn handle_dataset_lifecycle_deleted_message(
+        &self,
+        DatasetLifecycleMessageDeleted { dataset_id, .. }: &DatasetLifecycleMessageDeleted,
+    ) -> Result<(), InternalError> {
+        self.dataset_entry_repo
+            .delete_dataset_entry(dataset_id)
+            .await
+            .int_err()
+    }
+
+    async fn handle_dataset_lifecycle_renamed_message(
+        &self,
+        DatasetLifecycleMessageRenamed {
+            dataset_id,
+            new_dataset_name,
+            ..
+        }: &DatasetLifecycleMessageRenamed,
+    ) -> Result<(), InternalError> {
+        self.dataset_entry_repo
+            .update_dataset_entry_name(dataset_id, new_dataset_name)
+            .await
+            .int_err()
+    }
+
+    async fn has_datasets_indexed(&self) -> Result<bool, InternalError> {
+        let stored_dataset_entries_count = self
+            .dataset_entry_repo
+            .dataset_entries_count()
+            .await
+            .int_err()?;
+
+        Ok(stored_dataset_entries_count > 0)
+    }
+
+    async fn index_datasets(&self) -> Result<(), InternalError> {
+        use futures::TryStreamExt;
+
+        let dataset_handles: Vec<_> = self.dataset_repo.get_all_datasets().try_collect().await?;
+
+        self.concurrent_dataset_handles_processing(dataset_handles)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn concurrent_dataset_handles_processing(
+        &self,
+        dataset_handles: Vec<DatasetHandle>,
+    ) -> Result<(), InternalError> {
+        let mut join_set = tokio::task::JoinSet::new();
+        let now = self.time_source.now();
+
+        for dataset_handle in dataset_handles {
+            let task_account_repository = self.account_repository.clone();
+            let task_dataset_entry_repo = self.dataset_entry_repo.clone();
+            let task_now = now;
+
+            join_set.spawn(async move {
+                let owner_account_id = get_dataset_owner_id(
+                    &task_account_repository,
+                    &dataset_handle.alias.account_name,
+                )
+                .await?;
+                let dataset_entry = DatasetEntry::new(
+                    dataset_handle.id,
+                    owner_account_id.clone(),
+                    dataset_handle.alias.dataset_name,
+                    task_now,
+                );
+
+                task_dataset_entry_repo
+                    .save_dataset_entry(&dataset_entry)
+                    .await
+                    .int_err()
+            });
+        }
```

Comment:
Why do we need this to run in parallel? It's a one time job that runs at startup

Reply:
No urgent need, but it was not difficult to do (and most importantly fast)